### DISCUSSION TITLE
Dualshock 4 (don`t official) doesn`t connect!

### DISCUSSION THREAD

**Discussion by WowMan963** [ndeadly/MissionControl#775] [Q&A]
Hello!

I wanted to connect the second gamepad to the Nintendo switch, but for some reason it doesn't connect. The first dual shock 4 is from another manufacturer and is older, but connects without problems. The second dual shock 4 is not connected. I understand that most likely it's just that no official controller was supported.

who knows how you can make your own controller's support or how to get it?

**Comment by ndeadly** [comment]
Some unofficial controllers just don't work. They often pair with the console, but it disconnects them straight afterwards. I'm not sure why because I don't have one to do the necessary debugging. If you're tried resetting the controller and it's still not working, then this is likely what's going on.

**Reply by NuclleaR** [reply]
@ndeadly I's not strong in C++ to try debug, but how do you find and connect gamepad? maybe some device PID/VID or something? I have unofficial dualshock 4 which works greate when I connect it with USB sys-con

**Reply by ndeadly** [reply]
They're primarily detected via VID/PID matching, yeah. Detection is not the problem. These controllers do actually connect, but some internal error occurs within the bluetooth stack and they're disconnected at one of the lower layers before the system is made aware. The problem is that Nintendo's layer uses a generic code for failure that discards the actual internal error code, so the cause of the error can't be traced without a debugger and controller in hand. This is specifically a Bluetooth problem, so the fact it's working with sys-con is irrelevant.

**Reply by NuclleaR** [reply]
I see. How can I help to identify the controller? any info that I can provide from Windows ? or try with hid viewer?
Switch can't connect it in general.

**Reply by ndeadly** [reply]
As far as Mission Control is concerned the controller _is_ identified. It know's it's a PS4 controller, the bluetooth stack even opens a connection to it, but something fails during pairing that causes the stack to drop the connection. If that were to succeed then the controller would probably work without any additional changes. I can probably patch the binary where the error occurs, but without owning one of these controllers I can't locate where it fails. There are many such locations that generate an event with the same generic failure code. The issue is specific to the Switch bluetooth stack/driver, so the only way to get additional information on the failure is to debug the bluetooth module on the console using gdb while trying to pair the controller.

If you mean identify the controller as in identify whether it's one of the fakes or not, then I don't really have an answer for you there. I don't think that can be done reliably through software alone. You would have to follow one of the various online guides for visibly identifying fakes.

Sometimes though, even official controllers won't connect. In those cases you can resolve the issue by pressing the reset button on the underside of the controller.

**Reply by Masamune3210** [reply]
Given the absolute abominations I have seen over the years when it comes to third party controllers and their firmwares, it wouldn't surprise me at all if some of them just outright violate the bluetooth spec in places, might be what's peeving the stack in the switch off. Maybe someone could use a sniffer program and a computer to capture the data the controller sends?

**Reply by ndeadly** [reply]
I'm not sure if that would help. It might not be a violation of anything, and simply something the stack wasn't configured to support. And even if it were, I'm not sure my knowledge of the spec is complete enough to spot the issue. I would still need to debug with a controller in order to find the location to patch anyway.
